Sources/SwiftProtobuf/SetPathDecoder.swift

```diff
@@ -0,0 +1,343 @@
+// Sources/SwiftProtobuf/SetPathDecoder.swift - Path decoder (Setter)
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Decoder which sets value of a field by its path.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+/// Describes errors can occure during decoding a proto by path.
+public enum PathDecodingError: Error {
+
+  /// Describes a mismatch in type of the fields.
+  ///
+  /// If a value of type A is applied to a path with type B.
+  /// this error will be thrown.
+  case typeMismatch
+}
+
+extension Message {
+  static func number(for field: String) -> Int? {
+    guard let type = Self.self as? _ProtoNameProviding.Type else {
+      return nil
+    }
+    return type._protobuf_nameMap.number(forJSONName: field)
+  }
+}
+
+struct SetPathDecoder<T: Message>: Decoder {
+
+  private let path: String
+  private let value: Any?
+  private var number: Int?
+
+  init(path: String, value: Any?) {
+    self.path = path
+    self.value = value
+    if let firstPathComponent {
+      self.number = T.number(for: firstPathComponent)
+    }
+  }
+
+  var firstPathComponent: String? {
```

Comment:
Similar concerns about the paths, might be worth splitting once and tracking positions, etc.